File: scala/src/main/scala/bbc/dojonorth/lyrics/MusixMatchSongMatcher.scala
package bbc.dojonorth.lyrics

import scala.io.Source
import scala.language.postfixOps

/**
 * This class loads song titles and artists from the MSD matches file
 * Download the source data at: http://labrosa.ee.columbia.edu/millionsong/sites/default/files/AdditionalFiles/mxm_779k_matches.txt.zip
 */
class MusixMatchSongMatcher {

  def loadSongMetaData(filename: String): Map[String,MSDTrack] = {
    val trackLines = loadDataFile(filename)

    trackLines map { line: String =>
      val track = parseMsdTrackLine(line)
      (track.trackId -> track)
    } toMap
  }

  def loadDataFile(filename: String): Iterator[String] = {
    val allLines = Source.fromFile(filename).getLines()
    allLines withFilter { ln => !ln.startsWith("#") } withFilter { ln => ln.startsWith("TR") }
  }

  def parseMsdTrackLine(line: String): MSDTrack = {
    val splitLine = line.split("<SEP>")
    val trackId = splitLine(0)
    val artist = splitLine(1)
    val title = splitLine(2)
    val mxmId = splitLine(3) toInt

    MSDTrack(trackId, mxmId, artist, title)
  }
}

File: javascript/lib/lyrics/MusixMatchSongMatcher.js
var _ = require('lodash');
var fs = require('fs');

var MusixMatchSongMatcher = function MusixMatchSongMatcher () {};

MusixMatchSongMatcher.prototype.loadDataFile = function (filename) {
    var lines = fs.readFileSync(filename, {encoding: 'utf8'}).split('\n');

    return lines.filter(function(line) {
        return line.indexOf('#') !== 0 && line.indexOf('TR') !== -1;
    });
};

MusixMatchSongMatcher.prototype.parseMsdTrackLine = function (line) {
    var splitLine = line.split('<SEP>');

    return {
        trackId: splitLine[0],
        mxmId: parseInt(splitLine[3], 10),
        artist: splitLine[1],
        title: splitLine[2]
    }
};

MusixMatchSongMatcher.prototype.loadSongMetaData = function (filename) {
    var trackLines = this.loadDataFile(filename);
    var parsedTracks = {};

    _.each(trackLines, function (line) {
        var track = this.parseMsdTrackLine(line);
        parsedTracks[track.trackId] = track;
    }.bind(this));

    return parsedTracks;
};

module.exports = MusixMatchSongMatcher;